#Atividade Ativa de Lógica de Programação e Estrutura de Dados

###   Criação de um algoritmo que faça um conjunto de cinquenta numeros aleatórios para serem ordenados

*   Criação e embaralhamento dos números a biblioteca **random**;

*   Ordenação desse conjunto de valores com as funções **Bubble** e **Quick** *sort*;

*   Medição de tempo dessas funções;

*   Criação de um relatório explicando detalhadamente o algoritmo.


In [ ]:
#Instalação das Bibliotecas Sorting, Random e funções "List" e "Any" da Biblioteca Typing

!pip install sorting
import sorting
import random
from typing import List, Any

---
- Eu ultilizarei a biblioteca ***Sorting*** aqui apenas pela praticidade na construção do código, para construí-los com menos linhas;
- Igualmente com a ***Typing***, utilizando as funções List e Any, por costume e facilidade ao lidar com mais tipos de teste sem precisar declarar ou mudar o **tipo** de dado a todo momento.

In [ ]:
#Gerando cinquenta valores aleatórios de 333 a 666

numeros_aleatorio: List[Any] = random.sample(range(200, 800), 50)

if __name__ == "__main__":
    print("Bagunçado:")
    print(numeros_aleatorio)


Bagunçado:
[384, 202, 584, 729, 701, 358, 749, 743, 712, 250, 464, 338, 633, 412, 393, 468, 735, 401, 291, 304, 417, 366, 602, 526, 402, 225, 373, 452, 561, 321, 439, 516, 243, 636, 370, 652, 588, 457, 469, 539, 787, 593, 264, 343, 711, 428, 442, 766, 669, 285]


---
- Eu decidi por não usar o método "shuffle", mas criar uma amóstra aleatória desde o "range", com o "random.sample()";
- Decidi utilizar um range dos números 200 a 800 só para não ser parecido com nenhum outro código.

In [ ]:
#Realizando a ordenação pela função "bubble" da bibliotéca Sorting

%%time
ordenacao_bubble = sorting.bubble(numeros_aleatorio)
print("Ordenação por BUBBLESORT:")
print(ordenacao_bubble)
print("")

Ordenação por BUBBLESORT:
[202, 225, 243, 250, 264, 285, 291, 304, 321, 338, 343, 358, 366, 370, 373, 384, 393, 401, 402, 412, 417, 428, 439, 442, 452, 457, 464, 468, 469, 516, 526, 539, 561, 584, 588, 593, 602, 633, 636, 652, 669, 701, 711, 712, 729, 735, 743, 749, 766, 787]

CPU times: user 840 µs, sys: 0 ns, total: 840 µs
Wall time: 850 µs


In [ ]:
#Realizando a ordenação pela função "quick" da bibliotéca Sorting

%%time
ordenacao_quick = sorting.quick(numeros_aleatorio)
print("Ordenação por QUICKSORT:")
print(ordenacao_quick)
print("")

Ordenação por QUICKSORT:
[202, 225, 243, 250, 264, 285, 291, 304, 321, 338, 343, 358, 366, 370, 373, 384, 393, 401, 402, 412, 417, 428, 439, 442, 452, 457, 464, 468, 469, 516, 526, 539, 561, 584, 588, 593, 602, 633, 636, 652, 669, 701, 711, 712, 729, 735, 743, 749, 766, 787]

CPU times: user 1.61 ms, sys: 0 ns, total: 1.61 ms
Wall time: 1.62 ms


##**Conclusão**

- De acordo com os testes realizados com o algoritmo, a função **bubble** é mais rápida na ordenção de números aleatórios tendo o *wall time* de: 850 µs;
- Enquanto a função **quick** tem uma *wall time* de 1.62 ms.

Logo a função de ordenação ***bubble*** é a mais eficiente.